### PR TITLE
[PyTorch / Inference] Change Model server to Torchserve for PyTorch Inference for 1.6.0

### PR DESCRIPTION
**Pull request by dhanainme**
### Torchserve support for PyTorch Inference

Created after closing #347 

This is part of 2 PRs in sagemaker-pytorch-inference-toolkit / dlc to add support for TorchServe in Sagemaker.

Other PR  that needs to be merged before this PR :  https://github.com/aws/sagemaker-pytorch-inference-toolkit/pull/79

### Callouts 
* This PR has only inference image for 1.6.0. I had to create empty training files to ensure that i am able to build. 
* This PR depends on sagemaker-pytorch-inference>=2 which has not been released yet. This depends on [79 in sagemaker-pytorch-inference-toolkit](https://github.com/aws/sagemaker-pytorch-inference-toolkit/pull/79)

### Testing done

* Docker Image from sagemaker-pytorch-inference-toolkit was tested exhaustively to ensure the consumed packages - sagemaker-pytorch-inference-toolkit & sagemaker-inference-toolkit works. More details in https://github.com/aws/sagemaker-pytorch-inference-toolkit/pull/79 

## Checklist
- [x] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [build] | [test] | [build, test] | [ec2, ecs, eks, sagemaker]
- [x] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [x] (If applicable) I've documented below the tests I've run on the DLC image
- [] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ARG PYTHON_VERSION=3.6.10
+ARG TS_VERSION=0.1.1
+
+# See http://bugs.python.org/issue19846
+ENV LANG C.UTF-8
+ENV LD_LIBRARY_PATH /opt/conda/lib/:$LD_LIBRARY_PATH
+ENV PATH /opt/conda/bin:$PATH
+ENV SAGEMAKER_SERVING_MODULE sagemaker_pytorch_serving_container.serving:main
+ENV TEMP=/home/model-server/tmp
+
```
Comment (choidongyeon) [inline pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu:17]: Can you add the URL to our inference binary here?

```
ARG PT_INFERENCE_URL=https://aws-pytorch-binaries.s3-us-west-2.amazonaws.com/r1.6.0_inference/20200727-223446/b0251e7e070e57f34ee08ac59ab4710081b41918/cpu/torch-1.6.0-cp36-cp36m-manylinux1_x86_64.whl
```

File: pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+ && pip install torchserve==$TS_VERSION \
+ && pip install torch-model-archiver==$TS_VERSION
+
+
+ # Uninstall and re-install torch and torchvision from the PyTorch website
+RUN pip uninstall -y torch \
+ && pip install --no-cache-dir -U https://download.pytorch.org/whl/cu102/torch-1.6.0-cp36-cp36m-linux_x86_64.whl \
+ && pip uninstall -y torchvision \
+ && pip install --no-deps --no-cache-dir -U \
+    https://download.pytorch.org/whl/cpu/torchvision-0.7.0%2Bcpu-cp36-cp36m-linux_x86_64.whl
```
Comment (choidongyeon) [inline pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu:76]: Is there any particular reason why you're using the CUDA 10.2 binary (vs CPU) here?

Could you replace these lines with:

```
RUN pip uninstall -y torch \
 && pip install --no-cache-dir -U ${PT_INFERENCE_URL} \
 && pip uninstall -y torchvision \
 && pip install --no-deps --no-cache-dir -U \
    https://torchvision-build.s3.amazonaws.com/1.6.0/cpu/torchvision-0.7.0-cp36-cp36m-manylinux1_x86_64.whl
```

File: pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ARG PYTHON_VERSION=3.6.10
+ARG TS_VERSION=0.1.1
+
+# See http://bugs.python.org/issue19846
+ENV LANG C.UTF-8
+ENV LD_LIBRARY_PATH /opt/conda/lib/:$LD_LIBRARY_PATH
+ENV PATH /opt/conda/bin:$PATH
+ENV SAGEMAKER_SERVING_MODULE sagemaker_pytorch_serving_container.serving:main
+ENV TEMP=/home/model-server/tmp
+
```
Comment (choidongyeon) [inline pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: Similarly, the GPU binary URL will be:

```
ARG PT_INFERENCE_URL=https://aws-pytorch-binaries.s3-us-west-2.amazonaws.com/r1.6.0_inference/20200727-223446/b0251e7e070e57f34ee08ac59ab4710081b41918/gpu/torch-1.6.0-cp36-cp36m-manylinux1_x86_64.whl
```

Comment (dhanainme) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: Do you have a similar URL for Torchvision ?

Comment (choidongyeon) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: Yes, for gpu it is:
```
https://torchvision-build.s3.amazonaws.com/1.6.0/cpu/torchvision-0.7.0-cp36-cp36m-manylinux1_x86_64.whl
```

Comment (dhanainme) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: This URL is not accesible. Just confirming

Comment (choidongyeon) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: addressed offline :)

File: pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu (hunk 1 above)
Comment (dhanainme) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.cpu:17]: Done

File: pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu (hunk 3 above)
Comment (dhanainme) [reply pytorch/inference/docker/1.6.0/py3/Dockerfile.gpu:18]: Done

**Comment by dhanainme** [comment]
Just a quick followup. Can this me merged if it looks good.

**Comment by choidongyeon** [comment]
@dhanainme, can you confirm that torchvision is working properly as discussed offline?

File: pytorch/buildspec.yml [hunk 4]
```diff
@@ -23,6 +23,16 @@
...
+        source: docker/build_artifacts/torchserve-entrypoint.py
+        target: torchserve-entrypoint.py
+      config:
+        source: docker/build_artifacts/config.properties
+        target: config.properties
+      deep_learning_container:
+        source: docker/build_artifacts/deep_learning_container.py
+        target: deep_learning_container.py
 
   images:
```
Comment (TusharKanekiDey) [inline pytorch/buildspec.yml:37]: Need to add BuildCPUPTInferencePy3DockerImage and BuildGPUPTInferencePy3DockerImage as well to build Inference Images.

**Comment by YYStreet** [comment]
Close in favor of https://github.com/aws/deep-learning-containers/pull/482

**Pull request closed without merging**